cpu/esp32/periph/i2c_hw.c

```diff
@@ -288,23 +288,49 @@ int i2c_release(i2c_t dev)
     return 0;
 }
 
-#define _i2c_return_on_error(dev) \
+#define _i2c_return_on_error_read(dev) \
     if (_i2c_bus[dev].results & I2C_ARBITRATION_LOST_INT_ENA) { \
-        LOG_TAG_ERROR("i2c", "arbitration lost dev=%u\n", dev); \
+        LOG_TAG_DEBUG("i2c", "arbitration lost dev=%u\n", dev); \
         _i2c_reset_hw (dev); \
-__asm__ volatile ("isync"); \
+        __asm__ volatile ("isync"); \
         return -EAGAIN; \
     } \
     else if (_i2c_bus[dev].results & I2C_ACK_ERR_INT_ENA) { \
-        LOG_TAG_ERROR("i2c", "ack error dev=%u\n", dev); \
+        LOG_TAG_DEBUG("i2c", "ack error dev=%u\n", dev); \
         _i2c_reset_hw (dev); \
-__asm__ volatile ("isync"); \
-        return -EIO; \
+        __asm__ volatile ("isync"); \
+        return -ENXIO; \
     } \
     else if (_i2c_bus[dev].results & I2C_TIME_OUT_INT_ENA) { \
-        LOG_TAG_ERROR("i2c", "bus timeout dev=%u\n", dev); \
+        LOG_TAG_DEBUG("i2c", "bus timeout dev=%u\n", dev); \
         _i2c_reset_hw (dev); \
-__asm__ volatile ("isync"); \
+        __asm__ volatile ("isync"); \
+        return -ETIMEDOUT; \
+    }
+
+#define _i2c_return_on_error_write(dev) \
+    if (_i2c_bus[dev].results & I2C_ARBITRATION_LOST_INT_ENA) { \
+        LOG_TAG_DEBUG("i2c", "arbitration lost dev=%u\n", dev); \
+        _i2c_reset_hw (dev); \
+        __asm__ volatile ("isync"); \
+        return -EAGAIN; \
+    } \
+    else if (_i2c_bus[dev].results & I2C_ACK_ERR_INT_ENA) { \
+        LOG_TAG_DEBUG("i2c", "ack error dev=%u\n", dev); \
+        uint32_t _tx_fifo_cnt = _i2c_hw[dev].regs->status_reg.tx_fifo_cnt; \
+        _i2c_reset_hw (dev); \
+        __asm__ volatile ("isync"); \
+        if (_tx_fifo_cnt >= len) { \
```

Comment:
It would be good if you added a comment to this macro explaining the reasoning behind this.
Just a short note that this is for discriminating data and address NACKs, so you don't have to look up error codes to understand the macro :wink:

Reply:
@benpicco I tried my best.